misk/src/main/kotlin/misk/cloud/gcp/storage/LocalStorageRpc.kt

```diff
@@ -0,0 +1,582 @@
+        return try {
+            // NB(mmihic): We have to lock both the source and target in deterministic order
+            // regardless of whether they are being used as the source or target, otherwise
+            // we risk running into deadlocks
+            val (source, target) = request.source to request.target
+            val (lock1, lock2) = arrayOf(source.blobId, target.blobId)
+                    .sortedWith(Comparator { b1, b2 -> b1.compareTo(b2) })
+
+            withWriteLock(lock1) {
+                withWriteLock(lock2) {
+                    val sourceMetadata = getMetadataForReading(source.blobId, request.sourceOptions)
+                            ?: throw StorageException(404, "${source.blobId.fullName} not found")
+                    val sourceContentFile =
+                            contentRoot.resolve(source.blobId.toPath(sourceMetadata.generation))
+
+                    val existingTargetMetadata =
+                            getMetadataForWriting(target.blobId, request.targetOptions)
+                    val newTargetMetadata =
+                            request.target.nextGenerationMetadata(existingTargetMetadata)
+                    val targetContentFile =
+                            contentRoot.resolve(target.blobId.toPath(newTargetMetadata.generation))
+
+                    // Copy from source to a temporary file, then atomically move from the
+                    // temp file into the target target
+                    val tempFile = createTempUploadFile(target.blobId)
+                    Files.copy(sourceContentFile, tempFile, REPLACE_EXISTING)
+                    Files.move(tempFile, targetContentFile, ATOMIC_MOVE, REPLACE_EXISTING)
+                    writeMetadata(target.blobId, newTargetMetadata)
+
+                    existingTargetMetadata?.let {
+                        val oldTargetContentPath =
+                                contentRoot.resolve(target.blobId.toPath(it.generation))
+                        try {
+                            Files.deleteIfExists(oldTargetContentPath)
+                        } catch (_: IOException) {
+                            // This is ok
+                        }
+                    }
+
+                    val sourceContentSize = getContentSize(sourceContentFile)
+                    StorageRpc.RewriteResponse(
+                            request,
+                            request.target,
+                            sourceContentSize,
+                            true,
+                            "token",
+                            sourceContentSize)
+                }
+            }
+        } catch (e: IOException) {
+            throw StorageException(e)
+        }
+    }
+
+    private fun writeMetadata(blobId: BlobId, metadata: BlobMetadata) {
+        // NB(mmihic): Must be called with a write lock
+        val metadataJson = metadataAdapter.toJson(metadata).toByteArray()
```

Comment:
ByteString.encodeUtf8() instead? I worked on mobile stuff that shipped everywhere and I’m still recovering from the default charset being relevant for calls like this

Reply:
I think the case is a bit different - `toByteArray` explicitly uses UTF-8 as its default parameter.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-byte-array.html